A fix to make full blocks able to act as ladders

D-prog-55 opened the pull request:
### Purpose
____________________
This Pull Request aims to "fix" something that can be quite frustrating: the fact that full blocks can't act as ladders. I'm making this PR because, while helping a friend making a scaffolding block (which essentially acts as a full block ladder), the player was not climbing the block. After further testing, we realised that the check for whever it's a ladder doesn't work when the block's bounding block hits the edge of the block "grid". The only way I could "fix" it was by adding a version of the `Block#isLadder` method that accepts a EnumFacing argument.

### Implementation
____________________
This PR, as stated above, adds a version of the `Block#isLadder` method that takes an extra `EnumFacing` argument at the end. But I've left the original one, it's still used in `ForgeHooks$isLivingOnLadder` (the new one is called as a fallback in case the "non side-able" one returns false). The only problem is that some blocks that override `Block#isLadder` might have some unexpected behaviours, but that won't be a problem if the block overrides the `BlockLadder` (or `BlockVine`) class, as they're using the "side-able" one, meaning that vanilla ladders for exemple won't allow you to climb them from the thin side (which is probably the reason why full block ladders were not a thing in the first place). And a "side effect" of this is that we can make blocks that act as ladders on some sides, but not all.
And of course, the EnumFacing will always be an horizontal one, it will never be either `EnumFacing.UP` or `EnumFacing.DOWN`.

### Demonstration
____________________
This PR comes with a test mod that adds one block (its id is `test_full_ladder:the_ladder`), that has a full bounding box and that properly acts as a ladder. The block doesn't have a model/texture, or an item, but I don't think they're really needed.

Commit:
Full block ladder fix

1) Added a side sensitive version of Block#isLadder
2) Kept (and deprecated) the "old" Block#isLadder
3) ForgeHooks$isLivingOnLadder uses the "new" version as a fallback is the "old" returns false
4) ForgeHooks$isLivingOnLadder now also checks the block on the side for ladder (making full blocks able to act as ladders)
5) Fixed BlockLadder and BlockVine classes to reflect the above points
6) Added a test mod with one block (test_full_ladder:the_ladder) to demonstrate the point n°4)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=5051) <br/>All committers have signed the CLA.

MajorTuvok commented:
I've a question, wouldn't it be easier to call through to the old function in the new one with the additional parameter and swap all uses of it to the new one?
(I haven't looked at your changes yet, so I'm propably wrong)

MajorTuvok commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -642,17 +642,19 @@ public static double getPlayerVisibilityDistance(EntityPlayer player, double xzD
-            return state.getBlock().isLadder(state, world, pos, entity);
+        	Block b1 = state.getBlock();
+        	Block b2 = world.getBlockState(pos.offset(side)).getBlock();
+        	return b1.isLadder(state, world, pos, entity) || b1.isLadder(state, world, pos, entity, side) || b2.isLadder(state, world, pos, entity) || b2.isLadder(state, world, pos, entity, side);
```
The main difference lies here as far as I can tell, this would save you from calling an deprecated function and remove 2 checks

MajorTuvok commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -642,17 +642,19 @@ public static double getPlayerVisibilityDistance(EntityPlayer player, double xzD
         return value >= maxXZDistance ? maxXZDistance : value;
     }
 
-    public static boolean isLivingOnLadder(@Nonnull IBlockState state, @Nonnull World world, @Nonnull BlockPos pos, @Nonnull EntityLivingBase entity)
```
Removing this breaks binary compatability...
Deprecate and call through to the new one with an null side argument?

MajorTuvok commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -664,7 +666,7 @@ public static boolean isLivingOnLadder(@Nonnull IBlockState state, @Nonnull Worl
                         BlockPos tmp = new BlockPos(x2, y2, z2);
                         state = world.getBlockState(tmp);
-                        if (state.getBlock().isLadder(state, world, tmp, entity))
+                        if (state.getBlock().isLadder(state, world, tmp, entity, side))
```
Ok, so here are you only calling the new implementation - you either have to make your new implementation call through to the old method, or use both in order to check

Commit:
Fixed my mistakes.

Sorry about that.

strikerrocker commented:
Follow this for test mods https://github.com/MinecraftForge/MinecraftForge/wiki/If-you-want-to-contribute-to-Forge

strikerrocker commented:
Follow this guidelines for the test mod https://github.com/MinecraftForge/MinecraftForge/wiki/Guidelines-for-Writing-a-Test-Mod

temp1011 commented on patches/minecraft/net/minecraft/block/BlockLadder.java.patch:
```diff
@@ -1,10 +1,19 @@
+ import net.minecraft.util.math.BlockPos;
+ import net.minecraft.world.IBlockAccess;
+ import net.minecraft.world.World;
++import net.minecraftforge.client.model.b3d.B3DModel.Face;
```
import in patch file

MajorTuvok commented:
In order to remain compatible with existing implementations I still think you should call through to the original Method in Block and simply allow Mods to override the side sensitive Version

Commit:
Fixed more mistakes

- The "old" ForgeHooks$isLivingOnLadder is re-added (to not cause binary compatability break)
- The "new" ForgeHooks$isLivingOnLadder isn't directly calling the deprecated Block#isLadder anymore
- The "new" Block#isLadder is by default returning the "old" one's return
- Various \t turned into "    " and other line savings
- Using the RegistryEvent for the test mod

Commit:
Fixed more more mistakes

- Added license in test mod
- Removed the B3D import

MajorTuvok commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -642,17 +642,24 @@ public static double getPlayerVisibilityDistance(EntityPlayer player, double xzD
     {
         boolean isSpectator = (entity instanceof EntityPlayer && ((EntityPlayer)entity).isSpectator());
         if (isSpectator) return false;
+        if (side == null) return false;
```
Might eventually need to run old behaviour if side is null, cause people might expect that to happen (it would be possible to allow Side to passed in as null, but that will introduce different problems) ... 
Though I don't think it is that commonly used as that one would really need to keep behaviour at that level, and spam everything with nullability checks... 
What do you guys think?

LexManos replied:
You need to have a @Nullable annotation on the param if you accept null. You also short circuit a lot of hooks that call the old version. So... making the binary compat useless.

liach commented on src/test/java/net/minecraftforge/debug/block/FullBlockLadderTest.java:
```diff
@@ -0,0 +1,65 @@
+    @EventHandler
+    public static void onPreInit(FMLPreInitializationEvent e)
+    {
+        MinecraftForge.EVENT_BUS.register(BlockRegistry.class);
```
You should add a way to disable this test mod .

mezz commented:
How does the scaffolding from other mods work?
IC2, Openblocks, and Factory Tech have "Scaffolding", Immersive Engineering has "Steel Scaffolding" and MFR has "Vine Scaffolding".

loordgek commented:
what you are telling that you cant make Scaffold is untrue 
https://github.com/dalapo/FactoryTech/blob/master/src/main/java/dalapo/factech/block/BlockScaffold.java#L33

but i think the PR is still good in that in it adds a facing to it

oliviathevampire commented:
@mezz they work with help of making the model/bounding box smaller than a full block

loordgek commented:
Factory Tech is using full blocks

oliviathevampire commented:
@loordgek Maybe they are having a smaller bounding box

loordgek commented:
i cant find it

LexManos commented on patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:
```diff
@@ -151,7 +151,31 @@
++            {
++                return true;
++            }
++        }
```
This patch is horribly huge.

LexManos commented on patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:
```diff
@@ -151,7 +151,31 @@
+         int j = MathHelper.func_76128_c(this.func_174813_aQ().field_72338_b);
+         int k = MathHelper.func_76128_c(this.field_70161_v);
+-
++        
```
useless

LexManos commented on patches/minecraft/net/minecraft/block/BlockVine.java.patch:
```diff
@@ -32,7 +32,7 @@
      }
 +    /*************************FORGE START***********************************/
-+    @Override public boolean isLadder(IBlockState state, IBlockAccess world, BlockPos pos, EntityLivingBase entity){ return true; }
++    @Override public boolean isLadder(IBlockState state, IBlockAccess world, BlockPos pos, EntityLivingBase entity, EnumFacing side){ return (state.func_177229_b(field_176273_b) && side == EnumFacing.SOUTH) || (state.func_177229_b(field_176279_N) && side == EnumFacing.NORTH) || (state.func_177229_b(field_176278_M) && side == EnumFacing.WEST) || (state.func_177229_b(field_176280_O) && side == Enu ...
```
The sides shouldnt matter for vines or ladders, as their bounding box is less then a full block.

LexManos commented on patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:
```diff
@@ -247,15 +271,40 @@
++            else
++            {
++                side = p_191986_1_ > 0 ? side.func_176735_f() : side.func_176746_e();
++            }
```
Again, stupidly large patch file, and whitespace changes.

LexManos commented on patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch:
```diff
@@ -451,7 +451,28 @@
++            {
++                side = p_191986_1_ > 0 ? side.func_176735_f() : side.func_176746_e();
++            }
++        }
```
Duplicate code from above patch, as weel as massivly large patch.

LexManos commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -642,17 +642,24 @@ public static double getPlayerVisibilityDistance(EntityPlayer player, double xzD
         if (!ForgeModContainer.fullBoundingBoxLadders)
         {
-            return state.getBlock().isLadder(state, world, pos, entity);
+        	return state.getBlock().isLadder(state, world, pos, entity, side) || world.getBlockState(pos.offset(side)).getBlock().isLadder(state, world, pos, entity, side);
```
What if your size is less then a block, or you're not actually hitting that side block.

LexManos reviewed (commented):
At the end of the day this PR is poor on a lot of cases. And as people have said there are ways to make this work without these breaking changes.
